### ISSUE TITLE
Mangling for C++ pack indexing

### ISSUE THREAD

**Issue by cor3ntin** [itanium-cxx-abi/cxx-abi#175]
Hey folks

Pack indexing (https://wg21.link/p2662r3) was approved for C++26.
We will need to mangle these things.

A pack indexing is constituted of a pattern (which is either a type or an expression) and an indexing expression.
I am unfortunately not familiar enough with the ABI to offer suggestions

**Comment by rjmccall** [comment]
Looks like it's a new type-level operator and a new expression-level operator.  Apparently it will eventually be accepted as a template-name as well, but that's not currently included?  Anyway, this should be straightforward; we just need to add new forms to `<expression>` and `<type>`.

Typical uses of this are probably going to exacerbate our existing problems with partial substitution, though.  Consider when a pack indexing operator is used in the signature of a templated member of a class template and we can't expand the pack in the instantiated dependent signature of the template:

```c++
template <class... T> struct A {
  template <unsigned I> T...[I] get();
```

We'll need to solve that problem.

**Comment by cor3ntin** [comment]
Ping on that, it would be nice to settle on a grammar! Thanks

**Comment by jicama** [comment]
> ```c++
> template <class... T> struct tuple {
>   template <unsigned I> T...[I] get();
> };
> ```

This seems analogous to

```c++
template <class... T> struct A {
  template <int I> void f(T (*...p)[I]) {}
};

int main()
{
  A<int,double,char>().f<2>(0,0,0);
}
```

where we don't attempt to represent the partial instantiation in the mangled name, we represent the individual expanded parameters (as PAT__iPAT__dPAT__c).  Similarly, I'd think we can just represent the selected type as the return type of get().

**Comment by jicama** [comment]
> ```c++
> template <class... T> struct tuple {
>   template <unsigned I> T...[I] get();
> };
> ```

I would expect mangling of T...[I] to just refer to the two template parameters, following the mangling for template parameters at different depths from PR #85 .

**Comment by zygoloid** [comment]
I agree that using a reference to a template parameter by depth and index makes sense here. The scheme in #85 counts levels inwards from the first non-substituted level, and so `L` in this case would be -1, for which we don't yet have a mangling. Maybe something like `TM<1 - L>_<index - 1>_`, so `I` in the previous example would be `TM0__`?

One additional thought on this:

The "count from the first non-substituted level" rule means that we get different template parameter numbering in constraints versus the rest of the signature (due to the different number of substituted levels), which is annoying for demanglers to deal with. It'd be nice to change that to also use this new negative-levels numbering relative to the total number of substituted levels for the `<encoding>`, so that in

```c++
template<typename T, typename U> concept C = true;
template<typename ...T> struct A {
    template<typename ...U, int I> void f(T...[I], U...[I]) requires C<T...[I], U...[I]>;
};
```

... we'd use `TM0__` and `T_` to refer to `T` and `U` in both the constraint and the parameter types, rather than instead using `T_` and `TL0__` for `T` and `U` in the constraint as we currently do. Obviously that's an ABI break, and maybe it's already too late, but I think it would be a substantial improvement to the ease of demangling.

**Comment by cor3ntin** [comment]
We'll also need something for https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2026/p3670r4.pdf - I'll try to see if i can come up with something
